addons/account/models/account_move.py

```diff
@@ -3210,78 +3207,141 @@ def _get_integrity_hash_fields(self):
+        for journal, journal_moves in self.grouped('journal_id').items():
+            chains = journal_moves.grouped(lambda m: (m.move_type, m.sequence_prefix)).items()
+            for chain, chain_moves in chains:
+                last_move_in_chain = chain_moves.sorted('sequence_number')[-1]
+                if not self._is_move_restricted(last_move_in_chain, force_hash=force_hash):
+                    continue
+
+                common_domain = [
+                    ('journal_id', '=', journal.id),
+                    ('sequence_prefix', '=', last_move_in_chain.sequence_prefix),
+                ]
+                last_move_hashed = self.env['account.move'].search([
+                    *common_domain,
+                    ('inalterable_hash', '!=', False),
+                ], order='sequence_number desc', limit=1)
+
+                domain = self.env['account.move']._get_move_hash_domain([
+                    *common_domain,
+                    ('sequence_number', '<=', last_move_in_chain.sequence_number),
+                    ('inalterable_hash', '=', False),
+                    ('date', '>', last_move_in_chain.company_id._get_user_fiscal_lock_date())
+                ], force_hash=True)
+                if last_move_hashed and not include_pre_last_hash:
+                    # Hash moves only after the last hashed move, not the ones that may have been posted before the journal was set on restrict mode
+                    domain.extend([('sequence_number', '>', last_move_hashed.sequence_number)])
+
+                # On the accounting dashboard, we are only interested on whether there are documents to hash or not
+                # so we can stop the computation early if we find at least one document to hash
+                if early_stop and self.env['account.move'].sudo().search_count(domain):
+                    return True
+                else:
+                    moves_to_hash = self.env['account.move'].sudo().search(domain, order='sequence_number')
+                if not moves_to_hash and last_move_in_chain and force_hash and raise_if_no_document:
+                    if len(self) == 1:
+                        msg = _("This document should not be hashed because a later document is already hashed.")
+                    else:
+                        msg = _("These documents should not be hashed because a later document is already hashed.")
+                    raise UserError(msg)
+                if raise_if_gap and moves_to_hash and moves_to_hash[0].sequence_number + len(moves_to_hash) - 1 != moves_to_hash[-1].sequence_number:
+                    raise UserError(
+                        _("An error occurred when computing the inalterability. A gap has been detected in the sequence.")
+                    )
+
+                # If the last hashed move is a move pre-v4, but the new one is a v4 move
+                # then do not use the hash of the last move as a base for the new one
```

Comment:
Why?
If we don't, we will have 2 branches of hash without knowing when/why the branching started